dart/dynamics/ShapeFrame.cpp

```diff
@@ -0,0 +1,380 @@
+}
+
+//==============================================================================
+ConstShapePtr ShapeFrame::getShape() const
+{
+  return mShapeFrameP.mShape;
+}
+
+//==============================================================================
+VisualAddon* ShapeFrame::getVisualAddon(const bool createIfNull)
+{
+  VisualAddon* visualData = getVisualAddon();
+
+  if (createIfNull && nullptr == visualData)
+    return createVisualAddon();
+
+  return visualData;
+}
+
+//==============================================================================
+CollisionAddon* ShapeFrame::getCollisionAddon(const bool createIfNull)
+{
+  CollisionAddon* collisionData = getCollisionAddon();
+
+  if (createIfNull && nullptr == collisionData)
+    return createCollisionAddon();
+
+  return collisionData;
+}
+
+//==============================================================================
+DynamicsAddon* ShapeFrame::getDynamicsAddon(const bool createIfNull)
+{
+  DynamicsAddon* dynamicsData = getDynamicsAddon();
+
+  if (createIfNull && nullptr == dynamicsData)
+    return createDynamicsAddon();
+
+  return dynamicsData;
+}
+
+//==============================================================================
+void ShapeFrame::draw(renderer::RenderInterface* ri,
+                     const Eigen::Vector4d& color,
+                     bool useDefaultColor) const
+{
+  auto visualData = getVisualAddon();
+
+  if (!visualData || visualData->isHidden())
+    return;
+
+  ri->pushMatrix();
+  ri->transform(getRelativeTransform());
+
+  if (useDefaultColor)
+    mShapeFrameP.mShape->draw(ri, visualData->getRGBA());
+  else
+    mShapeFrameP.mShape->draw(ri, color);
+
+  ri->popMatrix();
+}
+
+//==============================================================================
+size_t ShapeFrame::incrementVersion()
+{
+  return ++mShapeFrameP.mVersion;
+}
+
+//==============================================================================
+size_t ShapeFrame::getVersion() const
+{
+  return mShapeFrameP.mVersion;
+}
+
+//==============================================================================
+ShapeFrame::ShapeFrame(Frame* parent, const Properties& properties)
+  : common::AddonManager(),
+    Entity(ConstructFrame),
+    Frame(parent, ""),
+    mShapeUpdatedSignal(),
+    mRelativeTransformUpdatedSignal(),
+    onShapeUpdated(mShapeUpdatedSignal),
+    onRelativeTransformUpdated(mRelativeTransformUpdatedSignal)
+{
+  setProperties(properties);
+}
+
+//==============================================================================
+ShapeFrame::ShapeFrame(Frame* parent,
+                       const std::string& name,
+                       const ShapePtr& shape)
+  : common::AddonManager(),
+    Entity(ConstructFrame),
+    Frame(parent, name),
+    mShapeUpdatedSignal(),
+    mRelativeTransformUpdatedSignal(),
+    onShapeUpdated(mShapeUpdatedSignal),
+    onRelativeTransformUpdated(mRelativeTransformUpdatedSignal)
+{
+  setName(name);
```

Comment:
I'm somewhat concerned about calling this here since ``setName`` is a virtual function, and this is a pure abstract class wherein the most-derived implementer of ``setName`` is not likely to be constructed yet.

I actually intend for ``Entity`` to become a completely abstract interface class, so in the future there will be no resolution at all for ``setName`` in this constructor.